compiler/rustc_traits/src/implied_outlives_bounds.rs

```diff
@@ -65,76 +64,101 @@ fn compute_implied_outlives_bounds<'tcx>(
+                        //   the cache, but hence without the `T: Trait<Foo=_1>`
+                        //   constraint. As a result, `_1` never gets resolved,
+                        //   and we get an ICE (in dropck).
+                        //
+                        // Therefore, we register any predicates involving
+                        // inference variables. We restrict ourselves to those
+                        // involving inference variables both for efficiency and
+                        // to avoids duplicate errors that otherwise show up.
+                        if obligation.has_infer_types_or_consts() {
+                            fulfill_cx.register_predicate_obligation(infcx, obligation.clone());
+                        }
+                        true
+                    }
                     ty::PredicateKind::WellFormed(arg) => {
                         wf_args.push(arg);
-                        vec![]
+                        true
                     }
 
                     ty::PredicateKind::RegionOutlives(ty::OutlivesPredicate(r_a, r_b)) => {
-                        vec![OutlivesBound::RegionSubRegion(r_b, r_a)]
+                        if first {
+                            false
+                        } else {
+                            implied_bounds.push(OutlivesBound::RegionSubRegion(r_b, r_a));
+                            true
+                        }
                     }
 
                     ty::PredicateKind::TypeOutlives(ty::OutlivesPredicate(ty_a, r_b)) => {
-                        let ty_a = infcx.resolve_vars_if_possible(ty_a);
-                        let mut components = smallvec![];
-                        push_outlives_components(tcx, ty_a, &mut components);
-                        implied_bounds_from_components(r_b, components)
+                        if first {
+                            false
+                        } else {
+                            let ty_a = infcx.resolve_vars_if_possible(ty_a);
+                            let mut components = smallvec![];
+                            push_outlives_components(tcx, ty_a, &mut components);
+                            let bounds = implied_bounds_from_components(r_b, components);
+                            implied_bounds.extend(bounds);
+                            true
+                        }
                     }
-                },
+                }
+            });
+
+            if !first {
+                break;
             }
-        }));
+
+            if first {
+                let _ = fulfill_cx.select_where_possible(infcx);
+            }
+
+            first = false;
+        }
     }
 
     // Ensure that those obligations that we had to solve
     // get solved *here*.
     match fulfill_cx.select_all_or_error(infcx).as_slice() {
         [] => Ok(implied_bounds),
-        _ => Err(NoSolution),
+        errors => bug!("{:?}", errors),
```

Comment:
I realized I left this in. But it's okay, because I want to do a crater run on this anyways.

(I want to make this change separately, anyways.)

Reply:
Okay, so this change breaks the `hyper` benchmark in rustc-perf.